packages/site-middleware/src/withSearchIndex.ts

```diff
@@ -63,47 +30,53 @@ export const withSearchIndex: MosaicMiddleware<SearchIndexSlice> = async (
   const matches = resolvedUrl.match(/(.*)[!/]/);
   const urlPath = matches?.length ? matches[1] : '';
 
-  const mosaicUrl = isSnapshotS3 || isSnapshotFile ? '' : res.getHeader('X-Mosaic-Content-Url');
-  const searchDataUrl = `${mosaicUrl}/search-data.json`;
-
   try {
-    let searchIndexData;
+    let searchIndex;
     if (isSnapshotFile) {
-      const env = getEnv(urlPath);
-      const { MOSAIC_SNAPSHOT_DIR: mosaicSnapshotDir } = env.data;
-      const filePath = path.join(process.cwd(), mosaicSnapshotDir, 'search-data.json');
-      const rawSearchIndexData = await loadLocalFile(filePath);
-      searchIndexData = JSON.parse(rawSearchIndexData);
+      const { snapshotDir } = getSnapshotFileConfig(urlPath);
+      const filePath = path.join(process.cwd(), snapshotDir, urlPath, 'search-data.json');
+      let fileExists = false;
+      try {
+        await fs.promises.stat(filePath);
+        fileExists = true;
+      } catch {}
+
+      if (fileExists) {
+        const rawSearchIndex = await loadLocalFile(filePath);
+        searchIndex = JSON.parse(rawSearchIndex);
+      }
     } else if (isSnapshotS3) {
-      const s3Key = `/search-data.json`.replace(/^\//, '');
-      const rawSearchIndexData = await loadS3File(s3Key);
-      searchIndexData = JSON.parse(rawSearchIndexData);
+      const s3Key = `${urlPath}/search-data.json`.replace(/^\//, '');
+      const { accessKeyId, bucket, region, secretAccessKey } = getSnapshotS3Config(s3Key);
+      const { keyExists, loadKey } = createS3Loader(region, accessKeyId, secretAccessKey);
+      const s3KeyExists = await keyExists(bucket, s3Key);
+      if (s3KeyExists) {
+        const rawSearchIndex = await loadKey(bucket, s3Key);
+        searchIndex = JSON.parse(rawSearchIndex);
+      }
     } else {
-      const response = await fetchSearchData(searchDataUrl);
-      if (response.ok && response.status !== 204) {
-        searchIndexData = await response.json();
-      } else {
-        const show500 = response.status !== 404 && response.status !== 204;
-        const show404 = response.status === 404;
-        let errorMessage = `Could not find any search data defined for ${searchDataUrl}`;
-        if (show500) {
-          errorMessage = `An un-expected error occurred reading ${searchDataUrl}`;
+      const mosaicUrl = res.getHeader('X-Mosaic-Content-Url');
+      const response = await fetch(`${mosaicUrl}${urlPath}/search-data.json`, {
```

Comment:
`urlPath` doesn't apply for search data. Unlike sharedConfig, `search-data.json` is created once at the project root, so this line should be 
```
const response = await fetch(`${mosaicUrl}/search-data.json`, {
```

The also applies to the snapshot/S3 namespaces. As written, they will all 404

Search-data is global for the entire site, and my plan was to implement namespace support within that single search-data blob (i.e. `searchData = {namespaceOne: [/*content*/],namespaceTwo: [/*content*/]}`)

It might actually make more sense to handle the namespaced search-data in separate files as proposed here, but that would mean changing the searchIndex plugin too.

Reply:
Pushed a fix and amended the tests.